calyx-opt/src/passes/compile_static.rs

```diff
@@ -346,6 +421,70 @@ impl Visitor for CompileStatic {
         Ok(Action::Change(Box::new(e)))
     }
 
+    fn start_while(
+        &mut self,
+        s: &mut ir::While,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        // if while body is static, then we want to make sure that the while
+        // body does not take the extra cycle incurred by the done condition
+        // So we replace the while loop with `enable` of a wrapper group
+        // that sets the go signal of the static group in the while loop body high
+        // (all static control should be compiled into static groups by
+        // `static_inliner` now). The done signal of the wrapper group should be
+        // the condition that the fsm of the while body is %0 and the port signal
+        // is 1'd0.
+        // For example, we replace
+        // wires {
+        // static group A<1> {
+        //     ...
+        //   }
+        //    ...
+        // }
+
+        // control {
+        //   while l.out {
+        //     A;
+        //   }
+        // }
+        // with
+        // wires {
+        //  group early_reset_A {
+        //     ...
+        //        }
+        //
+        // group while_wrapper_early_reset_A {
+        //       early_reset_A[go] = 1'd1;
+        //       while_wrapper_early_reset_A[done] = !l.out & fsm.out == 1'd0 ? 1'd1;
+        //     }
+        //   }
+        //   control {
+        //     while_wrapper_early_reset_A;
+        //   }
+        if s.cond.is_none() {
```

Comment:
This doesn't have to go in this PR, but I was thinking that we could probably support comb groups in this case as well (I was also thinking for supporting `static if` comb groups). I may be missing something, but it seems like all we would have to do is trigger the comb group assignments at cycle 0 (in this case add `fsm.out == 0` to each of the comb group assignments, and then inline them into the wrapper group). The comb assignments will be "done instantaneously" so the port will be updated to reflect the comb group assignments at cycle 0, which is what we want?

Reply:
I would be careful about `with` construct (see https://github.com/cucapra/calyx/discussions/927). Here is a particular case I ran into the last time I tried to handle `if`-`with` without creating extra groups out of it: https://github.com/cucapra/calyx/issues/911#issuecomment-1321237004

The tried and true way to see if your reasoning about `if`-`with` lowering being correct is trying to make it work with the polybench integration tests.